File: backend/models/Fee.js
const mongoose = require('mongoose');

const feeSchema = new mongoose.Schema({
  student: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  academicYear: {
    type: String,
    required: true
  },
  term: {
    type: String,
    enum: ['1st Term', '2nd Term', '3rd Term', 'Annual'],
    required: true
  },
  
  // Fee breakdown
  feeCategories: [{
    category: {
      type: String,
      required: true,
      enum: ['tuition', 'library', 'laboratory', 'sports', 'transport', 'lunch', 'examination', 'development', 'other']
    },
    description: String,
    amount: {
      type: Number,
      required: true,
      min: 0
    },
    dueDate: {
      type: Date,
      required: true
    },
    status: {
      type: String,
      enum: ['pending', 'paid', 'overdue', 'partial'],
      default: 'pending'
    },
    paidAmount: {
      type: Number,
      default: 0,
      min: 0
    },
    remainingAmount: {
      type: Number,
      default: function() { return this.amount; }
    }
  }],
  
  // Total calculations
  totalAmount: {
    type: Number,
    required: true,
    min: 0
  },
  totalPaid: {
    type: Number,
    default: 0,
    min: 0
  },
  totalPending: {
    type: Number,
    default: function() { return this.totalAmount; }
  },
  
  // Overall status
  overallStatus: {
    type: String,
    enum: ['pending', 'paid', 'overdue', 'partial'],
    default: 'pending'
  },
  
  // Payment history
  paymentHistory: [{
    transactionId: String,
    amount: {
      type: Number,
      required: true,
      min: 0
    },
    paymentMethod: {
      type: String,
      enum: ['cash', 'card', 'bank_transfer', 'online', 'cheque'],
      required: true
    },
    paymentDate: {
      type: Date,
      default: Date.now
    },
    receivedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    category: String, // which fee category this payment is for
    receiptNumber: String,
    notes: String,
    status: {
      type: String,
      enum: ['completed', 'pending', 'failed', 'refunded'],
      default: 'completed'
    }
  }],
  
  // Late fee information
  lateFee: {
    applicable: { type: Boolean, default: false },
    amount: { type: Number, default: 0 },
    appliedDate: Date
  },
  
  // Discount information
  discount: {
    type: { type: String, enum: ['percentage', 'fixed'] },
    value: { type: Number, default: 0 },
    reason: String,
    appliedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    appliedDate: Date
  },
  
  // Notifications sent
  notifications: [{
    type: {
      type: String,
      enum: ['reminder', 'overdue', 'payment_confirmation', 'late_fee']
    },
    sentDate: { type: Date, default: Date.now },
    method: [{ type: String, enum: ['email', 'sms', 'push'] }],
    message: String
  }],
  
  // Comments and notes
  notes: String,
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }
}, {
  timestamps: true
});

// Calculate totals before saving
feeSchema.pre('save', function(next) {
  this.totalPaid = this.paymentHistory.reduce((sum, payment) => {
    return payment.status === 'completed' ? sum + payment.amount : sum;
  }, 0);
  
  this.totalPending = this.totalAmount - this.totalPaid;
  
  // Update overall status
  if (this.totalPending <= 0) {
    this.overallStatus = 'paid';
  } else if (this.totalPaid > 0) {
    this.overallStatus = 'partial';
  } else {
    // Check if any fee is overdue
    const now = new Date();
    const hasOverdue = this.feeCategories.some(fee => 
      fee.dueDate < now && fee.status !== 'paid'
    );
    this.overallStatus = hasOverdue ? 'overdue' : 'pending';
  }
  
  // Update individual fee categories
  this.feeCategories.forEach(feeCategory => {
    const categoryPayments = this.paymentHistory.filter(payment => 
      payment.category === feeCategory.category && payment.status === 'completed'
    );
    
    feeCategory.paidAmount = categoryPayments.reduce((sum, payment) => sum + payment.amount, 0);
    feeCategory.remainingAmount = feeCategory.amount - feeCategory.paidAmount;
    
    if (feeCategory.remainingAmount <= 0) {
      feeCategory.status = 'paid';
    } else if (feeCategory.paidAmount > 0) {
      feeCategory.status = 'partial';
    } else {
      const now = new Date();
      feeCategory.status = feeCategory.dueDate < now ? 'overdue' : 'pending';
    }
  });
  
  next();
});

// Index for better query performance
feeSchema.index({ student: 1, academicYear: 1, term: 1 });
feeSchema.index({ overallStatus: 1 });
feeSchema.index({ 'feeCategories.dueDate': 1 });

module.exports = mongoose.model('Fee', feeSchema);

File: mobile_app/lib/core/models/fee.dart
class Fee {
  final String id;
  final String student;
  final String academicYear;
  final String term;
  final List<FeeCategory> feeCategories;
  final double totalAmount;
  final double totalPaid;
  final double totalPending;
  final String overallStatus;
  final List<PaymentHistory> paymentHistory;
  final LateFee? lateFee;
  final Discount? discount;
  final String? notes;
  final String createdBy;
  final String? updatedBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  Fee({
    required this.id,
    required this.student,
    required this.academicYear,
    required this.term,
    required this.feeCategories,
    required this.totalAmount,
    required this.totalPaid,
    required this.totalPending,
    required this.overallStatus,
    required this.paymentHistory,
    this.lateFee,
    this.discount,
    this.notes,
    required this.createdBy,
    this.updatedBy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Fee.fromJson(Map<String, dynamic> json) {
    return Fee(
      id: json['_id'],
      student: json['student'],
      academicYear: json['academicYear'],
      term: json['term'],
      feeCategories: (json['feeCategories'] as List)
          .map((e) => FeeCategory.fromJson(e))
          .toList(),
      totalAmount: json['totalAmount'].toDouble(),
      totalPaid: json['totalPaid'].toDouble(),
      totalPending: json['totalPending'].toDouble(),
      overallStatus: json['overallStatus'],
      paymentHistory: (json['paymentHistory'] as List)
          .map((e) => PaymentHistory.fromJson(e))
          .toList(),
      lateFee: json['lateFee'] != null ? LateFee.fromJson(json['lateFee']) : null,
      discount: json['discount'] != null ? Discount.fromJson(json['discount']) : null,
      notes: json['notes'],
      createdBy: json['createdBy'],
      updatedBy: json['updatedBy'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'student': student,
      'academicYear': academicYear,
      'term': term,
      'feeCategories': feeCategories.map((e) => e.toJson()).toList(),
      'totalAmount': totalAmount,
      'totalPaid': totalPaid,
      'totalPending': totalPending,
      'overallStatus': overallStatus,
      'paymentHistory': paymentHistory.map((e) => e.toJson()).toList(),
      'lateFee': lateFee?.toJson(),
      'discount': discount?.toJson(),
      'notes': notes,
      'createdBy': createdBy,
      'updatedBy': updatedBy,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class FeeCategory {
  final String category;
  final String? description;
  final double amount;
  final DateTime dueDate;
  final String status;
  final double paidAmount;
  final double remainingAmount;

  FeeCategory({
    required this.category,
    this.description,
    required this.amount,
    required this.dueDate,
    required this.status,
    required this.paidAmount,
    required this.remainingAmount,
  });

  factory FeeCategory.fromJson(Map<String, dynamic> json) {
    return FeeCategory(
      category: json['category'],
      description: json['description'],
      amount: json['amount'].toDouble(),
      dueDate: DateTime.parse(json['dueDate']),
      status: json['status'],
      paidAmount: json['paidAmount'].toDouble(),
      remainingAmount: json['remainingAmount'].toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'category': category,
      'description': description,
      'amount': amount,
      'dueDate': dueDate.toIso8601String(),
      'status': status,
      'paidAmount': paidAmount,
      'remainingAmount': remainingAmount,
    };
  }
}

class PaymentHistory {
  final String? transactionId;
  final double amount;
  final String paymentMethod;
  final DateTime paymentDate;
  final String? receivedBy;
  final String? category;
  final String? receiptNumber;
  final String? notes;
  final String status;

  PaymentHistory({
    this.transactionId,
    required this.amount,
    required this.paymentMethod,
    required this.paymentDate,
    this.receivedBy,
    this.category,
    this.receiptNumber,
    this.notes,
    required this.status,
  });

  factory PaymentHistory.fromJson(Map<String, dynamic> json) {
    return PaymentHistory(
      transactionId: json['transactionId'],
      amount: json['amount'].toDouble(),
      paymentMethod: json['paymentMethod'],
      paymentDate: DateTime.parse(json['paymentDate']),
      receivedBy: json['receivedBy'],
      category: json['category'],
      receiptNumber: json['receiptNumber'],
      notes: json['notes'],
      status: json['status'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'transactionId': transactionId,
      'amount': amount,
      'paymentMethod': paymentMethod,
      'paymentDate': paymentDate.toIso8601String(),
      'receivedBy': receivedBy,
      'category': category,
      'receiptNumber': receiptNumber,
      'notes': notes,
      'status': status,
    };
  }
}

class LateFee {
  final bool applicable;
  final double amount;
  final DateTime? appliedDate;

  LateFee({
    required this.applicable,
    required this.amount,
    this.appliedDate,
  });

  factory LateFee.fromJson(Map<String, dynamic> json) {
    return LateFee(
      applicable: json['applicable'],
      amount: json['amount'].toDouble(),
      appliedDate: json['appliedDate'] != null 
          ? DateTime.parse(json['appliedDate']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'applicable': applicable,
      'amount': amount,
      'appliedDate': appliedDate?.toIso8601String(),
    };
  }
}

class Discount {
  final String type;
  final double value;
  final String? reason;
  final String? appliedBy;
  final DateTime? appliedDate;

  Discount({
    required this.type,
    required this.value,
    this.reason,
    this.appliedBy,
    this.appliedDate,
  });

  factory Discount.fromJson(Map<String, dynamic> json) {
    return Discount(
      type: json['type'],
      value: json['value'].toDouble(),
      reason: json['reason'],
      appliedBy: json['appliedBy'],
      appliedDate: json['appliedDate'] != null 
          ? DateTime.parse(json['appliedDate']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'value': value,
      'reason': reason,
      'appliedBy': appliedBy,
      'appliedDate': appliedDate?.toIso8601String(),
    };
  }
}
